crates/qasm2/src/lex.rs

```diff
@@ -0,0 +1,772 @@
+    /// the next byte isn't in the valid range for OpenQASM 2, or if the file/stream has failed to
+    /// read into the buffer for some reason.
+    fn peek_byte(&mut self) -> PyResult<Option<u8>> {
+        if self.col >= self.line_buffer.len() && self.advance_line()? == 0 {
+            return Ok(None);
+        }
+        match self.line_buffer[self.col] {
+            b @ 0x80..=0xff => {
+                self.done = true;
+                Err(QASM2ParseError::new_err(message_generic(
+                    Some(&Position::new(&self.filename, self.line, self.col)),
+                    &format!("encountered a non-ASCII byte: {:02X?}", b),
+                )))
+            }
+            b => Ok(Some(b)),
+        }
+    }
+
+    /// Expect that the next byte is not a word continuation, providing a suitable error message if
+    /// it is.
+    fn expect_word_boundary(&mut self, after: &str, start_col: usize) -> PyResult<()> {
+        match self.peek_byte()? {
+            Some(c @ (b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9' | b'_')) => {
+                Err(QASM2ParseError::new_err(message_generic(
+                    Some(&Position::new(&self.filename, self.line, start_col)),
+                    &format!(
+                        "expected a word boundary after {}, but saw '{}'",
+                        after, c as char
+                    ),
+                )))
+            }
+            _ => Ok(()),
+        }
+    }
+
+    /// Complete the lexing of a floating-point value from the position of maybe accepting an
+    /// exponent.  The previous part of the token must be a valid stand-alone float, or the next
+    /// byte must already have been peeked and known to be `b'e' | b'E'`.
+    fn lex_float_exponent(&mut self, start_col: usize) -> PyResult<TokenType> {
+        if !matches!(self.peek_byte()?, Some(b'e' | b'E')) {
+            self.expect_word_boundary("a float", start_col)?;
+            return Ok(TokenType::Real);
+        }
+        // Consume the rest of the exponent.
+        self.next_byte()?;
+        if let Some(b'+' | b'-') = self.peek_byte()? {
+            self.next_byte()?;
+        }
+        // Exponents must have at least one digit in them.
+        if !matches!(self.peek_byte()?, Some(b'0'..=b'9')) {
+            return Err(QASM2ParseError::new_err(message_generic(
+                Some(&Position::new(&self.filename, self.line, start_col)),
+                "needed to see an integer exponent for this float",
+            )));
+        }
+        while let Some(b'0'..=b'9') = self.peek_byte()? {
+            self.next_byte()?;
+        }
+        self.expect_word_boundary("a float", start_col)?;
+        Ok(TokenType::Real)
+    }
+
+    /// Lex a numeric token completely.  This can return a successful integer or a real number; the
+    /// function distinguishes based on what it sees.
+    fn lex_numeric(&mut self, start_col: usize) -> PyResult<TokenType> {
+        let first = self.line_buffer[start_col];
```

Comment:
Should `first` be set to `peek_byte()` instead? It seems like these lex functions assume that `self.line_buffer[start_col]` will be the same as `self.peek_byte()`, no?

Reply:
No, the first byte has already been consumed by the main lexer loop (I think I was trying to avoid crazy repetition of `self.next_byte()?` in all the single-symbol cases).